Mathlib/RingTheory/LaurentSeries.lean

```diff
@@ -852,4 +876,302 @@ instance instLaurentSeriesComplete : CompleteSpace K⸨X⸩ :=
+  · set η : ℤₘ₀ˣ := Units.mk0 (Multiplicative.ofAdd f.order : Multiplicative ℤ) (coe_ne_zero)
+      with hη
+    obtain ⟨P, hP⟩ := exists_Polynomial_intValuation_lt F (η * γ)
+    use RatFunc.X ^ f.order * (P : RatFunc K)
+    have F_mul := f.ofPowerSeries_powerSeriesPart
+    obtain ⟨s, hs⟩ := Int.exists_eq_neg_ofNat (le_of_lt ord_nonpos)
+    rw [← hF, hs, neg_neg, ← ofPowerSeries_X_pow s, ← inv_mul_eq_iff_eq_mul₀] at F_mul
+    · erw [hs, ← F_mul, PowerSeries.coe_pow, PowerSeries.coe_X, RatFunc.coe_mul, zpow_neg,
+        zpow_ofNat, inv_eq_one_div (RatFunc.X ^ s), RatFunc.coe_div, RatFunc.coe_pow, RatFunc.coe_X,
+        RatFunc.coe_one, ← inv_eq_one_div, ← mul_sub, map_mul, map_inv₀, ← PowerSeries.coe_X,
+        valuation_X_pow, ← hs, ← RatFunc.coe_coe, ← coe_sub, ← coe_algebraMap,
+        valuation_of_algebraMap, ← Units.val_mk0
+        (a := ((Multiplicative.ofAdd f.order : Multiplicative ℤ) : ℤₘ₀)), ← hη]
+      apply inv_mul_lt_of_lt_mul₀
+      rwa [← Units.val_mul]
+    · simp only [PowerSeries.coe_pow, pow_ne_zero, PowerSeries.coe_X, ne_eq,
+        single_eq_zero_iff, one_ne_zero, not_false_iff]
+  · obtain ⟨s, hs⟩ := Int.exists_eq_neg_ofNat (Int.neg_nonpos_of_nonneg (not_lt.mp ord_nonpos))
+    have hf_coe : (PowerSeries.X ^ s * F : K⟦X⟧) = f := by
+      rw [← f.single_order_mul_powerSeriesPart, neg_inj.1 hs, hF, PowerSeries.coe_mul,
+        PowerSeries.coe_pow, PowerSeries.coe_X, single_pow, nsmul_eq_mul, mul_one, one_pow]
+    obtain ⟨P, hP⟩ := exists_Polynomial_intValuation_lt (PowerSeries.X ^ s * F) γ
+    use P
+    erw [← hf_coe, ← RatFunc.coe_coe, ← coe_sub, ← coe_algebraMap, valuation_of_algebraMap]
+    exact hP
+
+theorem coe_range_dense : DenseRange (Coe.coe : RatFunc K → K⸨X⸩) := by
+  rw [denseRange_iff_closure_range]
+  ext f
+  simp only [UniformSpace.mem_closure_iff_symm_ball, Set.mem_univ, iff_true, Set.Nonempty,
+    Set.mem_inter_iff, Set.mem_range, Set.mem_setOf_eq, exists_exists_eq_and]
+  intro V hV h_symm
+  rw [uniformity_eq_comap_neg_add_nhds_zero_swapped] at hV
+  obtain ⟨T, hT₀, hT₁⟩ := hV
+  obtain ⟨γ, hγ⟩ := Valued.mem_nhds_zero.mp hT₀
+  obtain ⟨P, _⟩ := exists_ratFunc_val_lt f γ
+  use P
+  apply hT₁
+  apply hγ
+  simpa only [add_comm, ← sub_eq_add_neg, gt_iff_lt, Set.mem_setOf_eq]
+
+end Dense
+
+section Comparison
+
+open RatFunc AbstractCompletion IsDedekindDomain.HeightOneSpectrum
+
+theorem inducing_coe : IsUniformInducing (Coe.coe : RatFunc K → K⸨X⸩) := by
+  rw [isUniformInducing_iff, Filter.comap]
+  ext S
+  simp only [exists_prop, Filter.mem_mk, Set.mem_setOf_eq, uniformity_eq_comap_nhds_zero,
+    Filter.mem_comap]
+  constructor
+  · rintro ⟨T, ⟨⟨R, ⟨hR, pre_R⟩⟩, pre_T⟩⟩
+    obtain ⟨d, hd⟩ := Valued.mem_nhds.mp hR
+    use {P : RatFunc K | Valued.v P < ↑d}
+    simp only [Valued.mem_nhds, sub_zero]
+    refine ⟨⟨d, by rfl⟩, subset_trans (fun _ _ ↦ pre_R ?_) pre_T⟩
+    apply hd
+    simp only [sub_zero, Set.mem_setOf_eq]
+    erw [← coe_sub, ← valuation_eq_LaurentSeries_valuation]
```

Comment:
Same as above.

Reply:
Nope, and I spent a while trying to understand it. I suspect it is the `coe` from `RatFunc` to `LaurentSeries`, but I will check.